### PR TITLE
Refactor contract development/deploy page

### PR DESCRIPTION
**Pull request by uF4No**
- Separates contract development and deployment in two different pages
- Adds compilers section
- Removes link to uncompleted page "L1 contract interface" in SDK/Tools
- Updates compiler version and hardhat config in all examples
- Adds Docker deprecation warning message

### CODE REVIEW

File: docs/dev/developer-guides/contracts/contracts.md [hunk 1]
```diff
@@ -31,7 +46,23 @@ Compilers are no longer released as Docker images and its usage is no longer rec
...
 - [hardhat-zksync-solc documentation](../../../api/hardhat/plugins.md#hardhat-zksync-solc)
 - [hardhat-zksync-vyper documentation](../../../api/hardhat/plugins.md#hardhat-zksync-vyper)
+
+## EVM compatibility
+
+Almost every smart contract written for EVM will be supported by zkSync 2.0 and will hold all key security invariants so that no additional security re-auditing will be required in most cases. A notable exception is the contracts that use the following EVM opcodes:
+
+- `SELFDESTRUCT`: it’s considered harmful and was deprecated in [EIP-6049](https://eips.ethereum.org/EIPS/eip-6049).
+- `EXTCODECOPY` - we've skip it for now because zkEVM opcodes are not identical to EVM ones anyway but it can be implemented if needed.
+- `CALLCODE` - it was deprecated on Ethereum in favor of `DELEGATECALL`.
```
Comment (krofax) [inline docs/dev/developer-guides/contracts/contracts.md:60]: Referencing where it was stated would be nice.
Link - https://eips.ethereum.org/EIPS/eip-2488

Comment (uF4No) [reply docs/dev/developer-guides/contracts/contracts.md]: solved

File: docs/dev/developer-guides/contracts/contracts.md [hunk 2]
```diff
@@ -31,7 +46,23 @@ Compilers are no longer released as Docker images and its usage is no longer rec
...
+**Learn more about how to install and configure these plugins in the links below:**
 
 - [hardhat-zksync-solc documentation](../../../api/hardhat/plugins.md#hardhat-zksync-solc)
 - [hardhat-zksync-vyper documentation](../../../api/hardhat/plugins.md#hardhat-zksync-vyper)
+
+## EVM compatibility
+
+Almost every smart contract written for EVM will be supported by zkSync 2.0 and will hold all key security invariants so that no additional security re-auditing will be required in most cases. A notable exception is the contracts that use the following EVM opcodes:
+
+- `SELFDESTRUCT`: it’s considered harmful and was deprecated in [EIP-6049](https://eips.ethereum.org/EIPS/eip-6049).
```
[the remark this answers is not in the document]
Comment (uF4No) [reply docs/dev/developer-guides/contracts/contracts.md]: where?!?!

[the remark this answers is not in the document]
Comment (uF4No) [reply docs/dev/developer-guides/contracts/contracts.md]: where!?!?

[the remark this answers is not in the document]
Comment (githubdoramon) [reply docs/dev/developer-guides/contracts/contracts.md]: "SELFDESTRUCT:" should be "SELFDESTRUCT - "

[the remark this answers is not in the document]
Comment (uF4No) [reply docs/dev/developer-guides/contracts/contracts.md]: 🙄 I though a "A" was missing 😅

[the remark this answers is not in the document]
Comment (githubdoramon) [reply docs/dev/developer-guides/contracts/contracts.md]: Yep.. the comment was not clear

File: docs/dev/developer-guides/contracts/contract-deployment.md [hunk 3]
```diff
@@ -0,0 +1,69 @@
+# Contract deployment
+
+To maintain the same security as in L1, the zkSync operator must publish on the Ethereum chain the contract code for each contract it deploys. However, if there are multiple contracts deployed with the same code, it will only publish it on Ethereum once. While deploying contracts for the first time may be relatively expensive, factories, which deploy contracts with the same code multiple times, can have huge savings compared to L1.
```
[the remark this answers is not in the document]
Comment (uF4No) [reply docs/dev/developer-guides/contracts/contract-deployment.md:3]: added a minor improvement, but I think it's better as is. There are no savings in L1, the fees are what they are, but factories will have savings in L2 compared to L1 due to the fact that the contract code will only be published once.

File: docs/dev/developer-guides/contracts/contract-deployment.md [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+We recommend using the [hardhat-zksync-deploy](../../../api/hardhat) plugin, to simplify the deployment process. It provides classes and methods to take care of all the deployment requirements, like generating the [bytecode hash of the contract](#format-of-bytecode-hash).
+
+Here's a [step-by-step guide on how to use it](../../../api/hardhat/getting-started.md).
+
+### Note on `factory deps`
+
+Under the hood, zkSync stores not bytecodes of contracts, but [specially formatted](#format-of-bytecode-hash) hashes of their bytecodes. You can see that even the [ContractDeployer](./system-contracts.md#contractdeployer) system contract accepts the bytecode hash of the deployed contract and not its bytecode. However, for contract deployment to succeed, the operator needs to know the bytecode. Exactly for this reason the `factory_deps` (i.e. factory dependencies) field for transactions is used: it contains the bytecodes that should be known to the operator for this transaction to succeed. Once the transaction succeeds, these bytecodes will be published on L1 and will be considered "known" to the operator forever.
```
[the remark this answers is not in the document]
Comment (uF4No) [reply docs/dev/developer-guides/contracts/contract-deployment.md]: fixed

File: docs/dev/developer-guides/contracts/contract-deployment.md [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+
+The 32-byte hash of the bytecode of a zkSync contract is calculated in the following way:
+
+- The first 2 bytes denote the version of bytecode hash format and are currently equal to `[1,0]`.
+- The second 2 bytes denote the length of the bytecode in 32-byte words.
+- The rest of the 28-byte (i.e. 28 low big-endian bytes) are equal to the last 28 bytes of the `sha256` hash of the contract's bytecode.
+
+### Differences in `CREATE` behaviour
+
+For the ease of supporting account abstraction, for each account, we split the nonce into two parts: _the deployment nonce_ and _the transaction nonce_. The deployment nonce is the number of contracts the account has deployed with `CREATE` opcode, while the transaction nonce is used for replay attack protection for the transactions.
```
[the remark this answers is not in the document]
Comment (uF4No) [reply docs/dev/developer-guides/contracts/contract-deployment.md:57]: added

**Review by krofax — APPROVED** [review]

**Pull request merged**